### PR TITLE
pass listener in integration test

### LINKED ISSUE #58936: We should pass listener to SecureServingOptions in test
As we have add `Listener` to `SecureServingOptions `, we should stop using function like `FindFreeLocalPort` in test cases and create a listener and pass to `SecureServingOptions `.

Because `FindFreeLocalPort` just create a tcp listener and get its port and close it. By this we can not make sure the port not be used between this listener close and the kube-apiserver listener start.

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**What this PR does / why we need it**:

pass listener to `SecureServingOptions` to prevent port in use flake.
partially fix #58936

**Special notes for your reviewer**:

/assign @liggitt @sttts @caesarxuchao 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: test/integration/etcd/etcd_storage_path_test.go [hunk 1]
```diff
@@ -705,22 +706,22 @@ func startRealMasterOrDie(t *testing.T, certDir string) (*allClient, clientv3.KV
...
+			listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
+			if err != nil {
+				t.Fatal(err)
+			}
+
 			kubeAPIServerOptions := options.NewServerRunOptions()
+
+			kubeAPIServerOptions.SecureServing.BindPort = port
 			kubeAPIServerOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
+			kubeAPIServerOptions.SecureServing.Listener = listener
```
Comment (sttts) [inline test/integration/etcd/etcd_storage_path_test.go]: :+1:

Can't we get rid of the whole retry loop with this? @enj

File: test/integration/examples/apiserver_test.go [hunk 2]
```diff
@@ -91,15 +92,15 @@ func TestAggregatedAPIServer(t *testing.T) {
 	kubeClientConfigValue := atomic.Value{}
 	go func() {
 		for {
-			// always get a fresh port in case something claimed the old one
-			kubePort, err := framework.FindFreeLocalPort()
+			listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: also get rid of the retry loop here.

File: test/integration/examples/apiserver_test.go [hunk 3]
```diff
@@ -154,9 +155,13 @@ func TestAggregatedAPIServer(t *testing.T) {
 			t.Log(err)
 			return false, nil
 		}
-		if _, err := kubeClient.Discovery().ServerVersion(); err != nil {
+
+		healthStatus := 0
+		kubeClient.Discovery().RESTClient().Get().AbsPath("/healthz").Do().StatusCode(&healthStatus)
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: why this change?

Comment (sttts) [reply test/integration/examples/apiserver_test.go]: Just because it is more strict? (as the listener will be active here already)

File: test/integration/examples/apiserver_test.go [hunk 4]
```diff
@@ -178,30 +183,32 @@ func TestAggregatedAPIServer(t *testing.T) {
...
-				"--cert-dir", wardleCertDir,
-				"--kubeconfig", kubeconfigFile.Name(),
-			})
-			if err := wardleCmd.Execute(); err != nil {
-				t.Log(err)
+			atomic.StoreInt32(wardlePort, int32(port))
+			options := sampleserver.NewWardleServerOptions(os.Stdout, os.Stderr)
+			options.Complete()
+
+			options.RecommendedOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: this is a step back. The flags are more abstract. Don't like to switch back to options. Can we come up with an hybrid of flags plus the listener in the options?

File: test/integration/examples/apiserver_test.go [hunk 5]
```diff
@@ -256,28 +264,34 @@ func TestAggregatedAPIServer(t *testing.T) {
...
-			if err := aggregatorCmd.Execute(); err != nil {
-				t.Log(err)
+			atomic.StoreInt32(aggregatorPort, int32(port))
+
+			options := kubeaggregatorserver.NewDefaultOptions(os.Stdout, os.Stderr)
+			if err := options.Complete(); err != nil {
+				t.Fatal(err)
+			}
+
+			options.RecommendedOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: same here, would prefer flags.

File: test/integration/examples/apiserver_test.go [hunk 6]
```diff
@@ -256,28 +264,34 @@ func TestAggregatedAPIServer(t *testing.T) {
 
 	go func() {
 		for {
-			// always get a fresh port in case something claimed the old one
-			aggregatorPortInt, err := framework.FindFreeLocalPort()
+
+			listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: also no loop anymore.

File: test/integration/etcd/etcd_storage_path_test.go (hunk 1 above)
Comment (hzxuzhonghu) [reply test/integration/etcd/etcd_storage_path_test.go]: Sure, I missed it.

File: test/integration/examples/apiserver_test.go (hunk 3 above)
Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Yes

File: test/integration/examples/apiserver_test.go (hunk 4 above)
Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Maybe write a fake `NewCommandStartWardleServer` here

File: test/integration/examples/apiserver_test.go (hunk 3 above)
Comment (sttts) [reply test/integration/examples/apiserver_test.go]: ok

File: test/integration/examples/apiserver_test.go (hunk 4 above)
Comment (sttts) [reply test/integration/examples/apiserver_test.go]: Can't we create the options, parse the flags and patch in the listener? (have not looked at the code whether it's feasible) The goal is to have as few custom plumbing code in the test as possible. I.e. as end-2-end as possible.

Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: I tried but failed, because in  `NewCommandStartWardleServer` flags is parsed and conflicts with flags parse in test. And BTW we have to create options and parse flags , overwrite cobra.Command.RunE. This way equals to rewrite a `NewCommandStartWardleServer`.

**Comment by hzxuzhonghu** [comment]
@sttts By rewriting `NewCommandStartAggregator` `NewCommandStartWarldServer`, I re-implement once, https://github.com/kubernetes/kubernetes/pull/58937/commits/281dfeda44a96a33a6c0df6f4cb2b7879eedcd4f. PTAL

**Comment by hzxuzhonghu** [comment]
BTW, I do not like this way, because it puts  Addflags in RunE.  
```            flags := cmd.Flags()
               o.RecommendedOptions.AddFlags(flags)
```

**Comment by hzxuzhonghu** [comment]
fix golint error, weird  golint error

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 7]
```diff
@@ -55,21 +55,23 @@ func NewCommandStartAggregator(out, err io.Writer, stopCh <-chan struct{}) *cobr
...
-			}
-			if err := o.Validate(args); err != nil {
-				return err
-			}
-			if err := o.RunAggregator(stopCh); err != nil {
-				return err
-			}
-			return nil
-		},
+	}
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: where is the difference to the old setting RunE in the struct construction above?

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: And why the comment?

File: staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go [hunk 8]
```diff
@@ -63,22 +63,23 @@ func NewCommandStartWardleServer(out, errOut io.Writer, stopCh <-chan struct{})
...
-			if err := o.Validate(args); err != nil {
-				return err
-			}
-			if err := o.RunWardleServer(stopCh); err != nil {
-				return err
-			}
-			return nil
-		},
 	}
+	// For integration test
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go]: same question here

File: test/integration/examples/apiserver_test.go [hunk 9]
```diff
@@ -184,24 +186,38 @@ func TestAggregatedAPIServer(t *testing.T) {
...
+			if err := o.Complete(); err != nil {
+				return err
+			}
+			if err := o.Validate(args); err != nil {
+				return err
+			}
+			err := o.RunWardleServer(stopCh)
+			return err
+		}
+		wardleCmd.SetArgs([]string{
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: please squash with the previous commit. I guess this change addresses my comments from before about flags?

File: test/integration/examples/apiserver_test.go [hunk 10]
```diff
@@ -184,24 +186,38 @@ func TestAggregatedAPIServer(t *testing.T) {
...
-		options.RecommendedOptions.CoreAPI.CoreAPIKubeconfigPath = kubeconfigFile.Name()
-
-		if err := options.RunWardleServer(stopCh); err != nil {
+
+		wardleCmd := sampleserver.NewCommandStartWardleServer(os.Stdout, os.Stderr, stopCh)
+		o := sampleserver.NewWardleServerOptions(os.Stdout, os.Stderr)
+		o.RecommendedOptions.SecureServing.Listener = listener
+		// parse flags
+		flags := wardleCmd.Flags()
+		o.RecommendedOptions.AddFlags(flags)
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: would it help to pass the initial `WardleServerOptions` (with a set listener) to `NewCommandStartWardleServer`? could avoid this acrobatic with flags and RunE.

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go (hunk 7 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: I move flag into RunE, otherwise if we parse flags in test, we will encounter flag redefined error.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: > And why the comment?

This is only for mock RunE, so we can pass listener into options

Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: Should we impl a fake one directly copy this code? I think this way is easy, and not change the original code.

File: test/integration/examples/apiserver_test.go (hunk 10 above)
Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: i can not find a way to pass to the initial `WardleServerOptions `. Because the `wardleCmd ` does not expose it.

Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Or we have another choice modify `NewCommandStartWardleServer` return the options.

Comment (sttts) [reply test/integration/examples/apiserver_test.go]: Why not pre-create it and inject into `NewCommandStartWardleServer`. Is the signature of `NewCommandStartWardleServer` fixed?

Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: We can't, because the `NewCommandStartWardleServer` does not take it as an argument.

File: test/integration/examples/apiserver_test.go (hunk 9 above)
Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Ok, after all comments addressed.

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 11]
```diff
@@ -49,9 +49,7 @@ type AggregatorOptions struct {
 }
 
 // NewCommandStartAggregator provides a CLI handler for 'start master' command
-func NewCommandStartAggregator(out, err io.Writer, stopCh <-chan struct{}) *cobra.Command {
-	o := NewDefaultOptions(out, err)
-
+func NewCommandStartAggregator(o *AggregatorOptions, stopCh <-chan struct{}) *cobra.Command {
```
Comment (enj) [inline staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: This does not look right (I do not think we normally pass in options to a command).

**Review by enj — COMMENTED** [review]
Changes to the storage path test look fine.  Not so sure about passing in options to a cobra command (just looks weird to me).

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go (hunk 11 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: We can set options only by this way.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: better call `o` defaults. I don't see anything wrong with this style.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: Indicate default options in comments to prevent rename `o` to `defaults` widespread ugly.

Comment (sttts) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: sgtm

File: test/integration/examples/apiserver_test.go [hunk 12]
```diff
@@ -90,52 +92,49 @@ func TestAggregatedAPIServer(t *testing.T) {
...
-				t.Log(err)
-			}
-			time.Sleep(100 * time.Millisecond)
+		listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		kubeAPIServerOptions := options.NewServerRunOptions()
+		kubeAPIServerOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: wasn't bind address and port automatic when the listener is set?

File: test/integration/examples/apiserver_test.go [hunk 13]
```diff
@@ -255,30 +258,30 @@ func TestAggregatedAPIServer(t *testing.T) {
...
+		if err != nil {
+			t.Fatal(err)
+		}
+		atomic.StoreInt32(aggregatorPort, int32(port))
+
+		o := kubeaggregatorserver.NewDefaultOptions(os.Stdout, os.Stderr)
+		o.RecommendedOptions.SecureServing.Listener = listener
+		aggregatorCmd := kubeaggregatorserver.NewCommandStartAggregator(o, stopCh)
+		aggregatorCmd.SetArgs([]string{
+			"--bind-address", "127.0.0.1",
```
Comment (sttts) [inline test/integration/examples/apiserver_test.go]: same here about address and port

File: test/integration/examples/apiserver_test.go (hunk 12 above)
Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Yes,  when listener is set.  BindAddress  and Port  no need to set. 
I set here to keep consistency.

Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: Actually we donot need `BindPort` and `BindAddress`.

Comment (hzxuzhonghu) [reply test/integration/examples/apiserver_test.go]: I see a nit bug, and will open a pr to fix it in `SecureServingOptions.ApplyTo`

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 14]
```diff
@@ -35,7 +35,8 @@ import (
 
 type SecureServingOptions struct {
 	BindAddress net.IP
-	BindPort    int
+	// BindPort is omitted when Listener is set, will serve https even with 0.
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: s/omitted/ignored/, maybe clearer.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 15]
```diff
@@ -160,7 +161,7 @@ func (s *SecureServingOptions) ApplyTo(c *server.Config) error {
 	if s == nil {
 		return nil
 	}
-	if s.BindPort <= 0 {
+	if s.BindPort <= 0 && s.Listener == nil {
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: :+1:

File: test/integration/etcd/etcd_storage_path_test.go [hunk 16]
```diff
@@ -704,47 +705,43 @@ func startRealMasterOrDie(t *testing.T, certDir string) (*allClient, clientv3.KV
...
-			tunneler, proxyTransport, err := app.CreateNodeDialer(kubeAPIServerOptions)
-			if err != nil {
-				t.Fatal(err)
-			}
-			kubeAPIServerConfig, sharedInformers, versionedInformers, _, _, err := app.CreateKubeAPIServerConfig(kubeAPIServerOptions, tunneler, proxyTransport)
-			if err != nil {
-				t.Fatal(err)
-			}
+		kubeAPIServerOptions.SecureServing.BindPort = port
+		kubeAPIServerOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```
Comment (sttts) [inline test/integration/etcd/etcd_storage_path_test.go]: with the last commit, these can go

Comment (hzxuzhonghu) [reply test/integration/etcd/etcd_storage_path_test.go]: ok

**Comment by hzxuzhonghu** [comment]
All comments addressed @sttts

**Comment by hzxuzhonghu** [comment]
@sttts @deads2k for approval.

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 17]
```diff
@@ -49,9 +49,8 @@ type AggregatorOptions struct {
 }
 
 // NewCommandStartAggregator provides a CLI handler for 'start master' command
-func NewCommandStartAggregator(out, err io.Writer, stopCh <-chan struct{}) *cobra.Command {
-	o := NewDefaultOptions(out, err)
-
+// with a default AggregatorOptions.
+func NewCommandStartAggregator(o *AggregatorOptions, stopCh <-chan struct{}) *cobra.Command {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: I still we should call them `defaults`. Makes it's purpose clearer.

**Comment by sttts** [comment]
Just the nit about calling the corba constructor arg `default`. Otherwise, lgtm.

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go (hunk 17 above)
Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: make sense

**Comment by hzxuzhonghu** [comment]
@sttts comments addressed

**Comment by hzxuzhonghu** [comment]
@deads2k please take another look

File: staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go [hunk 18]
```diff
@@ -49,27 +49,26 @@ type AggregatorOptions struct {
 }
 
 // NewCommandStartAggregator provides a CLI handler for 'start master' command
-func NewCommandStartAggregator(out, err io.Writer, stopCh <-chan struct{}) *cobra.Command {
-	o := NewDefaultOptions(out, err)
-
+// with a default AggregatorOptions.
+func NewCommandStartAggregator(defaults *AggregatorOptions, stopCh <-chan struct{}) *cobra.Command {
```
Comment (sttts) [inline staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: `o := *defaults`. Now, operating on defaults is strange.

Comment (hzxuzhonghu) [reply staging/src/k8s.io/kube-aggregator/pkg/cmd/server/start.go]: ic

**Comment by hzxuzhonghu** [comment]
@sttts done.

File: staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go [hunk 19]
```diff
@@ -57,9 +57,8 @@ func NewWardleServerOptions(out, errOut io.Writer) *WardleServerOptions {
 }
 
 // NewCommandStartWardleServer provides a CLI handler for 'start master' command
-func NewCommandStartWardleServer(out, errOut io.Writer, stopCh <-chan struct{}) *cobra.Command {
-	o := NewWardleServerOptions(out, errOut)
-
+// with a default WardleServerOptions.
+func NewCommandStartWardleServer(o *WardleServerOptions, stopCh <-chan struct{}) *cobra.Command {
 	cmd := &cobra.Command{
```
Comment (sttts) [inline staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go]: same here

Comment (hzxuzhonghu) [reply staging/src/k8s.io/sample-apiserver/pkg/cmd/server/start.go]: done
